src/validation_error.rs

```diff
@@ -0,0 +1,295 @@
+    }
+}
+
+#[derive(Serialize, Debug)]
+pub struct ParseErrorMetadata {
+    pub code: String,
+    pub title: String,
+    pub detail: String,
+    pub location: Option<ParseErrorLocation>,
+}
+
+impl From<serde_yaml::Error> for ParseErrorMetadata {
+    fn from(err: serde_yaml::Error) -> Self {
+        ParseErrorMetadata {
+            code: "parse_error".into(),
+            title: "Parse Error".into(),
+            detail: err.to_string(),
+            location: err.location().map(|location| location.into()),
+        }
+    }
+}
+
+#[derive(Serialize, Debug)]
+#[serde(untagged)]
+pub enum ValidationError {
+    // Schema Validation Errors
+    WrongType {
+        meta: ValidationErrorMetadata,
+    },
+    MultipleOf {
+        meta: ValidationErrorMetadata,
+    },
+    Maximum {
+        meta: ValidationErrorMetadata,
+    },
+    Minimum {
+        meta: ValidationErrorMetadata,
+    },
+    MaxLength {
+        meta: ValidationErrorMetadata,
+    },
+    MinLength {
+        meta: ValidationErrorMetadata,
+    },
+    Pattern {
+        meta: ValidationErrorMetadata,
+    },
+    MaxItems {
+        meta: ValidationErrorMetadata,
+    },
+    MinItems {
+        meta: ValidationErrorMetadata,
+    },
+    UniqueItems {
+        meta: ValidationErrorMetadata,
+    },
+    Items {
+        meta: ValidationErrorMetadata,
+    },
+    MaxProperties {
+        meta: ValidationErrorMetadata,
+    },
+    MinProperties {
+        meta: ValidationErrorMetadata,
+    },
+    Required {
+        meta: ValidationErrorMetadata,
+    },
+    Properties {
+        meta: ValidationErrorMetadata,
+    },
+    Enum {
+        meta: ValidationErrorMetadata,
+    },
+    AnyOf {
+        states: Vec<ValidationState>,
+        meta: ValidationErrorMetadata,
+    },
+    OneOf {
+        states: Vec<ValidationState>,
+        meta: ValidationErrorMetadata,
+    },
+    Const {
+        meta: ValidationErrorMetadata,
+    },
+    Contains {
+        meta: ValidationErrorMetadata,
+    },
+    ContainsMinMax {
+        meta: ValidationErrorMetadata,
+    },
+    Not {
+        meta: ValidationErrorMetadata,
+    },
+    DivergentDefaults {
+        meta: ValidationErrorMetadata,
+    },
+    Format {
+        meta: ValidationErrorMetadata,
+    },
+    Unevaluated {
+        meta: ValidationErrorMetadata,
+    },
+    Unknown {
+        meta: ValidationErrorMetadata,
+    },
+
+    // Other Validation Errors
+    UnresolvedJob {
+        meta: ValidationErrorMetadata,
+    },
+    #[cfg(not(feature = "js"))]
+    InvalidGlob {
+        meta: ValidationErrorMetadata,
+    },
+    #[cfg(not(feature = "js"))]
+    NoFilesMatchingGlob {
+        meta: ValidationErrorMetadata,
+    },
+
+    // Other Errors
+    Parse {
+        meta: ParseErrorMetadata,
+    },
+}
+
+impl From<&BoxedValicoError> for ValidationError {
+    fn from(err: &BoxedValicoError) -> Self {
+        if let Some(err) = err.downcast_ref::<valico::json_schema::errors::WrongType>() {
```

Comment:
Can this be done with a `match` somehow?  If you've tried and it can't be done, that's cool, just thought it seemed like the sort of thing that was suitable for a `match` rather than a big `else if` chain.

Reply:
I do agree that a match would be preferable here since we have no guarantee that 1) all errors are handled and 2) new errors introduced to valico will be handled in future.

However, I'm not aware of a way to refactor this into a match without first refactoring valico, since ValicoError is a trait rather than an enum.

The changes needed to valico would be a breaking change and a sizeable PR. I think it's probably worth raising an issue and seeing if we can get buy-in from the maintainer on this but I don't think it's worth holding out for that in this PR - but of course that's up to you! :)

Reply:
Yeah, no point holding up this PR for something like that.